Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp

```diff
@@ -118,21 +118,39 @@ void copy_data_block_bytes(ByteBuffer& to_block, u64 to_index, ByteBuffer& from_
     // 7. Return unused.
 }
 
-// 25.1.2.1 AllocateArrayBuffer ( constructor, byteLength ), https://tc39.es/ecma262/#sec-allocatearraybuffer
-ThrowCompletionOr<ArrayBuffer*> allocate_array_buffer(VM& vm, FunctionObject& constructor, size_t byte_length)
+// 25.1.2.1 AllocateArrayBuffer ( constructor, byteLength [ , maxByteLength ] ), https://tc39.es/proposal-resizablearraybuffer/#sec-allocatearraybuffer
+ThrowCompletionOr<ArrayBuffer*> allocate_array_buffer(VM& vm, FunctionObject& constructor, size_t byte_length, Optional<size_t> max_byte_length)
 {
-    // 1. Let obj be ? OrdinaryCreateFromConstructor(constructor, "%ArrayBuffer.prototype%", « [[ArrayBufferData]], [[ArrayBufferByteLength]], [[ArrayBufferDetachKey]] »).
+    // 1. Let slots be « [[ArrayBufferData]], [[ArrayBufferByteLength]], [[ArrayBufferDetachKey]] ».
+
+    // 2. If maxByteLength is present and not empty, then
+    // a. If byteLength > maxByteLength, throw a RangeError exception.
+    // b. Append [[ArrayBufferMaxByteLength]] to slots.
+    if (max_byte_length.has_value() && byte_length > max_byte_length.value())
+        return vm.throw_completion<RangeError>(ErrorType::InvalidByteLengthForResizableArrayBuffer, byte_length, max_byte_length.value());
+
+    // 3. Let obj be ? OrdinaryCreateFromConstructor(constructor, "%ArrayBuffer.prototype%", slots).
     auto obj = TRY(ordinary_create_from_constructor<ArrayBuffer>(vm, constructor, &Intrinsics::array_buffer_prototype, nullptr));
 
-    // 2. Let block be ? CreateByteDataBlock(byteLength).
+    // 4. Let block be ? CreateByteDataBlock(byteLength).
     auto block = TRY(create_byte_data_block(vm, byte_length));
 
-    // 3. Set obj.[[ArrayBufferData]] to block.
+    // 5. Set obj.[[ArrayBufferData]] to block.
     obj->set_buffer(move(block));
 
-    // 4. Set obj.[[ArrayBufferByteLength]] to byteLength.
+    // 6. Set obj.[[ArrayBufferByteLength]] to byteLength.
 
-    // 5. Return obj.
+    // 7. If maxByteLength is present and not empty, then
+    if (max_byte_length.has_value())
+        // a. If it is not possible to create a Data Block block consisting of maxByteLength bytes, throw a RangeError exception.
+        // b. NOTE: Resizable ArrayBuffers are designed to be implementable with in-place growth. Implementations reserve the right to throw if, for example, virtual memory cannot be reserved up front.
+        TRY_OR_THROW_OOM(vm, obj->buffer().try_ensure_capacity(max_byte_length.value()));
+
+    // NOTE: We always want to set this as max_byte_length is represented using an Optional in ArrayBuffer
```

Comment:
I don't think we need to always set this, the default construct of Optional is empty.